Objects/typeobject.c

```diff
@@ -3078,19 +3079,28 @@ type_new_descriptors(const type_new_ctx *ctx, PyTypeObject *type)
         }
     }
 
-    if (ctx->add_dict && ctx->base->tp_itemsize) {
-        type->tp_dictoffset = -(long)sizeof(PyObject *);
-        slotoffset += sizeof(PyObject *);
+    if (ctx->add_dict) {
+        if (ctx->base->tp_itemsize) {
+            type->tp_dictoffset = -(long)sizeof(PyObject *);
+            slotoffset += sizeof(PyObject *);
+        }
+        else {
+            assert((type->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0);
+            type->tp_flags |= Py_TPFLAGS_MANAGED_DICT;
+        }
     }
 
     if (ctx->add_weak) {
         assert(!ctx->base->tp_itemsize);
-        type->tp_weaklistoffset = slotoffset;
-        slotoffset += sizeof(PyObject *);
+        if (type->tp_flags & Py_TPFLAGS_MANAGED_DICT) {
+            type->tp_weaklistoffset = -4 * sizeof(PyObject *);
+        }
+        else {
+            type->tp_weaklistoffset = slotoffset;
+            slotoffset += sizeof(PyObject *);
+        }
     }
-    if (ctx->add_dict && ctx->base->tp_itemsize == 0) {
-        assert((type->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0);
-        type->tp_flags |= Py_TPFLAGS_MANAGED_DICT;
+    if (type->tp_flags & Py_TPFLAGS_MANAGED_DICT) {
```

Comment:
Likewise about `Py_TPFLAGS_MANAGED_DICT` implying something it didn't before (but only if the flag could possibly be set before `type_new_descriptors()` gets called).